### ISSUE TITLE
Any way to make logger optional?

### ISSUE THREAD

**Issue by shsteven** [danthorpe/Operations#144]
First of all, thank you @danthorpe  for creating this library!
It's even better than the original sample code thanks to all the improvements.

When using both Operations and a logging framework e.g. SwiftyBeaver, log.debug() stops working because both frameworks declare the same function.

Is there any way to make LoggerType optional?

**Comment by danthorpe** [comment]
Hi @shsteven - thanks a lot, much appreciated

In terms of disabling logging (when using `development` branch) - you can put `LogManager.enabled = false` - this will be made available in the next version, hopefully soon.

But, I don't think this is the issue, there is a name spacing clash here? Which is what I don't quite follow, however I don't think you're the only person to have this issue - and I'd really like to get to the bottom of it. Can you provide a code example of how you're using logging and what is going on?

`log` is a property on `Operation` - if you have a function called `log` then you can still use it inside an `Operation` subclass by specifying the module that it is defined in. E.g. 

```swift
class MyOperation: Operation {
    func execute() {
        log.info("This is an info message")
        SwiftyBeaver.log.debug("This is a log from another framework")
    }
}
```

On the `development` it's also possible to make the built in logger to pipe through to a custom logging framework. Although, this might be lacking in some features (e.g. severity). But essentially, you just need to set the `logger` property.

```swift
LogManager.logger = { message in
    SwiftBeaver.print(message)
}
```

I don't wish to drop the built in logger - because I actually find it really useful for developing and debugging the framework. Equally I can't really support every different external logging tool. So, I opted to make the cleanest separation, where the severity and queue was defined internally, but the block which actually processes the received message could be overridden.

One improvement I can make to this is to also pass the severity through to that block, which would allow for more flexibility is adapting it to an external system.

**Comment by danthorpe** [comment]
Also - I'm trying to avoid having any dependencies with Operations - hence why it doesn't use SwiftyBeaver itself.

**Comment by shsteven** [comment]
Hi @danthorpe ,

Thank you for your quick response!

Here's a concrete example:

```swift
import Operations
import SwiftyBeaver

class SampleOperation: Operation {
    override func execute() {
        log.info("This is an info message that works fine")
        log.debug("This doesn't work")
        SwiftyBeaver.debug("This works")
    }
}
```

Prefixing with framework name works.
And I completely agree seeing operation logs is very useful. 

However, it feels a bit awkward to use `SwiftyBeaver.debug()` instead of `log.debug()`. 
To avoid the name clash, I can setup SwityBeaver as `let logger = SwiftyBeaver.self` and use a different name `logger.debug()`. 
The downside is possible confusion for those who read the source code: am I suppose to use `log.debug()` or `logger.debug()`?

Perhaps there's a more elegant solution?

**Comment by shsteven** [comment]
An alternative could be renaming `operation.log` to `operation.logger`?

**Comment by danthorpe** [comment]
@shsteven appreciate that is a pragmatic solution.

But you could also define your `SwiftyBeaver` instance to be called `logger` instead of `log`. Also another logging framework might use `logger` and so there would be another clash. Modules names exist for this reason - to disambiguate symbols in mixed scope.

**Comment by shsteven** [comment]
@danthorpe yes, that's true.

It's a convention with the logging libraries. CocoaLumberjack, XCGLogger and SwiftyBeaver -- all of them suggest using `log.debug()` in their examples.
As a result, anyone who integrates with Operations and a logging library will scratch their head and wonder why `log.debug` stops working, ask the same question, and eventually go through the same debug process.

**Comment by danthorpe** [comment]
Well, I'm working on a ticket now to look at this more closely, and I'll add SwiftyBeaver to the example project and see what needs to be done to integrate it.

Partly my thoughts are - wouldn't it just be better to have the `Operation` logger send messages to the custom logger?

**Comment by shsteven** [comment]
> Partly my thoughts are - wouldn't it just be better to have the Operation logger send messages to the custom logger?

I think that'll work, to a degree.
The advantage of using a dedicated logging tool is the flexibility: logging to a file, colouring, showing line number and class name etc.
If we have the `Operation` logger wrapping around another logger, line number / class name is no longer relevant, for it will always point to the line where `Operation` logger is calling `SwiftyBeaver` logger.

**Comment by danthorpe** [comment]
Hmm, this is a good point. I'll get back to you with what I discover.

**Comment by shsteven** [comment]
Thanks!

**Comment by danthorpe** [comment]
Fixed in #145.
